tortoisebeacon/votescalc.go

```diff
@@ -0,0 +1,287 @@
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+)
+
+func (tb *TortoiseBeacon) calcVotesFromProposals(epoch types.EpochID) (votesFor, votesAgainst hashList) {
+	votesFor = make(hashList, 0)
+	votesAgainst = make(hashList, 0)
+
+	tb.timelyProposalsMu.RLock()
+
+	timelyProposals := tb.timelyProposals[epoch]
+
+	for p := range timelyProposals {
+		votesFor = append(votesFor, p)
+	}
+
+	tb.timelyProposalsMu.RUnlock()
+
+	tb.delayedProposalsMu.Lock()
+
+	delayedProposals := tb.delayedProposals[epoch]
+
+	for p := range delayedProposals {
+		votesAgainst = append(votesAgainst, p)
+	}
+
+	tb.delayedProposalsMu.Unlock()
+
+	tb.Log.With().Info("Calculated votes from proposals",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.String("for", fmt.Sprint(votesFor)),
+		log.String("against", fmt.Sprint(votesAgainst)))
+
+	return votesFor, votesAgainst
+}
+
+func (tb *TortoiseBeacon) calcVotesDelta(epoch types.EpochID, round types.RoundID) (forDiff, againstDiff hashList) {
+	tb.votesMu.Lock()
+	defer tb.votesMu.Unlock()
+
+	votesCount := tb.countFirstRoundVotes(epoch)
+
+	tb.Log.With().Info("Calculated first round votes",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("incomingVotes", fmt.Sprint(tb.incomingVotes[epochRoundPair{EpochID: epoch, Round: 1}])),
+		log.String("votesCount", fmt.Sprint(votesCount)))
+
+	ownFirstRoundVotes := tb.calcOwnFirstRoundVotes(epoch, votesCount)
+
+	tb.Log.With().Info("Calculated own first round votes",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesCount", fmt.Sprint(votesCount)),
+		log.String("ownFirstRoundVotes", fmt.Sprint(ownFirstRoundVotes)))
+
+	tb.calcVotesCount(epoch, round, votesCount)
+
+	tb.Log.With().Info("Calculated votes count",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesCount", fmt.Sprint(votesCount)))
+
+	ownCurrentRoundVotes := tb.calcOwnCurrentRoundVotes(epoch, round, votesCount)
+
+	tb.Log.With().Info("Calculated votes for one round",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("for", fmt.Sprint(ownCurrentRoundVotes.VotesFor)),
+		log.String("against", fmt.Sprint(ownCurrentRoundVotes.VotesAgainst)))
+
+	votesFor, votesAgainst := ownCurrentRoundVotes.Diff(ownFirstRoundVotes)
+
+	tb.Log.With().Info("Calculated votes diff for one round",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("for", fmt.Sprint(votesFor)),
+		log.String("against", fmt.Sprint(votesAgainst)))
+
+	return votesFor, votesAgainst
+}
+
+func (tb *TortoiseBeacon) countFirstRoundVotes(epoch types.EpochID) votesCountMap {
+	firstRoundInThisEpoch := epochRoundPair{
+		EpochID: epoch,
+		Round:   1,
+	}
+
+	tb.votesCache[firstRoundInThisEpoch] = make(map[p2pcrypto.PublicKey]votesSetPair)
```

Comment:
this need to be RLocked as well, unless there is a global lock - at which case please add a comment about it

Reply:
It's locked by `tb.votesMu` in `calcVotesDelta`, I've added a comment